metaflow/plugins/kfp/kfp_s3_sensor.py

```diff
@@ -7,28 +7,68 @@
 from email.policy import default
 import click
 
+import os
 import pathlib
 from typing import Dict
+import boto3
+import botocore
+import base64
+import json
+import marshal
+import time
+from urllib.parse import urlparse
+
+from typing import Tuple
+
+
+def get_elapsed_time_s3_bucket_and_key(
+    flow_name: str, kfp_run_id: str
+) -> Tuple[str, str]:
+    s3_path = os.path.join(
+        os.getenv("METAFLOW_DATASTORE_SYSROOT_S3"),
+        flow_name,
+        kfp_run_id,
+        "_s3_sensor",
+    )
+    s3_path_parsed = urlparse(s3_path)
+    bucket, key = s3_path_parsed.netloc, s3_path_parsed.path.lstrip("/")
+    return bucket, key
+
+
+def read_elapsed_time_s3_path(flow_name: str, kfp_run_id: str) -> float:
+    bucket, key = get_elapsed_time_s3_bucket_and_key(flow_name, kfp_run_id)
+    s3: botocore.client.S3 = boto3.client("s3")
+
+    try:
+        s3.head_object(Bucket=bucket, Key=key)
+    except botocore.exceptions.ClientError as e:
+        elapsed_time: float = 0.0
+    else:
+        s3_object: dict = s3.get_object(Bucket=bucket, Key=key)
+        elapsed_time: float = float(s3_object["Body"].read().decode("utf-8"))
+    return elapsed_time
+
+
+def write_elapsed_time_s3_path(
+    flow_name: str, kfp_run_id: str, elapsed_time: float
+) -> None:
+    bucket, key = get_elapsed_time_s3_bucket_and_key(flow_name, kfp_run_id)
+    s3: botocore.client.S3 = boto3.client("s3")
+    elapsed_time_binary_data = str(elapsed_time).encode("ascii")
+    s3.put_object(Body=elapsed_time_binary_data, Bucket=bucket, Key=key)
```

Comment:
does this have retry on failure?
why not use the Metaflow S3 client instead of your own boto3 call and client?

Reply:
My understanding was the Metaflow S3 client was for large files, and in this case, we have a small file. I used `boto3` to remain consistent with the way we do [code packaging](https://github.com/zillow/metaflow/blob/feature/kfp/metaflow/metaflow_environment.py#L85). According to the `boto3` docs, it comes with default retry support, specifically the [Legacy retry mode](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#legacy-retry-mode).

Reply:
can use this `get_aws_client`

https://github.com/zillow/metaflow/blob/0bc327e68432dc99342afe1594d27f0c4c992bb8/metaflow/plugins/aws/aws_client.py#L4-L4